app/controller.go

```diff
@@ -133,3 +155,98 @@ func startController(c *cli.Context) error {
 	})
 	return http.ListenAndServe(controlListener, router)
 }
+
+func checkPrerequisites(c *controller.Controller, replicas []types.Replica) error {
+	c.Lock()
+	if c.IsSnapDeletionInProgress {
+		c.Unlock()
+		return fmt.Errorf("Snapshot deletion process is in progress, %s is getting deleted", c.SnapshotName)
+	}
+
+	rf := c.ReplicationFactor
+	c.Unlock()
+	if len(replicas) != rf {
+		return fmt.Errorf("Can not remove a snapshot because, RF: %v, replica count: %v", rf, len(replicas))
+	}
+
+	for _, rep := range replicas {
+		r := rep // pin it
+		if err := c.IsReplicaRW(&r); err != nil {
+			return fmt.Errorf("Can't delete snapshot, err: %v", err)
+		}
+	}
+	return nil
+}
+
+func listSnapshot(c *controller.Controller) ([]string, error) {
+	c.Lock()
+	replicas := c.ListReplicas()
+	// checking RF here to ensure that required no of replicas
+	// are connected to the controller. We are checking it again
+	// since there is a gap after checkPrerequisites where replicas
+	// may get disconnected to controller.
+	rf := c.ReplicationFactor
```

Comment:
still the gap remains, its just narrowed down.

Reply:
donw